### ISSUE TITLE
Unhandled JS Exception when one of the item in object is a React-Native element

### ISSUE THREAD

**Issue by alexmngn** [KyleAMathews/deepmerge#76]
I have 2 objects I'd like to deep merge. One element in the object is a React-Native element.

First object:
```
const obj1 = {
  headerStyle: {
    color: 'blue',
  },
  headerLeft: (
    <Text>Back</Text>
  )
};

const obj2 = {
  headerStyle: {
    backgroundColor: 'red',
  },
  headerLeft: (
    <Text>Close</Text>
  )
}

merge(obj1, obj2);

```

I get the following error when I try to merge those 2 objects:
> Unhandled JS Exception: Maximum call stack size exceeded

![screen shot 2017-09-19 at 16 12 09](https://user-images.githubusercontent.com/4203845/30599739-5508c8da-9d55-11e7-87d7-6b08714dc68d.png)

**Comment by TehShrike** [comment]
I'm guessing that the React/React Native objects have cycles somewhere in their properties.

I assume that you are expecting the React Native objects to not be merged?

Can you find out if there's a canonical way to detect the React Native object type?  If so, we can add it to [is-mergeable-object](https://github.com/TehShrike/is-mergeable-object), which would fix the issue in your case.

**Comment by alexmngn** [comment]
@TehShrike I'm not expecting to merge the React Native objects but to get the last one indeed.
I cannot find a way to distinguish a React element of a React Native element

Both objects are quite similar:

ReactJS:
![screen shot 2017-09-19 at 16 55 44](https://user-images.githubusercontent.com/4203845/30602161-acd90736-9d5b-11e7-8bec-0eed23bb0b64.png)

React Native:
![screen_shot_2017-09-19_at_16_50_25](https://user-images.githubusercontent.com/4203845/30602177-bbc295a0-9d5b-11e7-83b3-cfcd9ab868a9.png)

One has a defined `_source_` but this doesn't seem to be accessible via code.

Although I would think the same rule should apply and we shouldn't merge ReactJS element either.

We can create a check for `$$typeof === Symbol(react.element)`

**Comment by TehShrike** [comment]
Yeah, this looks relevant: https://github.com/facebook/react/blob/b5ac963fb791d1298e7f396236383bc955f916c1/src/isomorphic/classic/element/ReactElement.js#L375-L381

**Comment by TehShrike** [comment]
Check over https://github.com/TehShrike/is-mergeable-object/pull/1

**Comment by TehShrike** [comment]
Any thoughts @alexmngn?

**Comment by macdja38** [comment]
It would be interesting to support merging circular references since in theory it would be possible.

**Comment by alexmngn** [comment]
@TehShrike Looks good to me 👍

**Comment by TehShrike** [comment]
Published as 1.5.2.
